Further UI components

m-mohr opened the pull request:
- Add UI component for Batch Job Results and tables for jobs, services and files

m-mohr commented:
I've also added tables for list_services, list_jobs and list_files.

Example:
![image](https://user-images.githubusercontent.com/8262166/109338039-00c77100-7866-11eb-8acb-3bc81ce35c2a.png)

m-mohr commented:
@soxofaan Do you have an idea why this happens?

![image](https://user-images.githubusercontent.com/8262166/109340035-c9a68f00-7868-11eb-9ae4-3923c7e88208.png)

m-mohr commented:
Additionally implemented functionality that allows iterating through collection items (e.g. pre-computed ARD).

@soxofaan Could you have a first look at whether this is all good from a Python/library perspective or whether you'd like me to change things? The SCRIPT_URL will get valid in the next days. 

Example:
![image](https://user-images.githubusercontent.com/8262166/110009247-0b2db300-7d1d-11eb-8a88-895afb25bc7f.png)

m-mohr commented:
We would also need to download assets from those API endpoints, but I think I'd aim for a more generic implementation that can also be re-used in downloadResults etc. - see #184

jdries commented on openeo/rest/connection.py:
```diff
@@ -588,6 +617,30 @@ def describe_collection(self, name) -> dict:
+    def collection_items(self, name, spatial_extent: Optional[Dict[str, float]] = None, temporal_extent: Optional[List[Union[str, datetime.datetime]]] = None) -> Paginate:
+        """
+        Loads items for a specific image collection.
+        May not be available for all collections.
```
It's actually not available for most collections, should we add a clean error in that case?

m-mohr replied:
I don't know, I think I followed the route of the rest of the python client where it does not do any "feature-checking" against e.g. the endpoints in "/". But happy to adopt whethever you prefer.

jdries replied:
Good point ;-)
But you are allowed to do feature checking, or otherwise to catch exceptions. In this case, I guess the server will return a 404 if item metadata is not available, so we could catch that, and translate it into a message that explains to the user that the feature is optional.

m-mohr replied:
The request is only done in the (hopefully generic) Paginate iterator so I can't really make that specific unfortunately. There's no request being sent in collection_items. So I could only check against the endpoints in /. But that is something that we could do for all endpoints in general, that doesn't seem too specific for this PR.

soxofaan commented on openeo/rest/connection.py:
```diff
@@ -55,6 +55,34 @@ def __init__(self, http_status_code: int = None,
+        response = self.con.get(self.nextUrl).json()
+        if isinstance(response['links'], list):
+            nextLink = next((link for link in response['links'] if link['rel'] == 'next'), None)
+            self.nextUrl = nextLink['href']
```
will fail if nextLink is None (there was no next link)

m-mohr replied:
Fixed

soxofaan commented on openeo/rest/connection.py:
```diff
@@ -55,6 +55,34 @@ def __init__(self, http_status_code: int = None,
         self.url = url
         super().__init__("[{s}] {c}: {m}".format(s=self.http_status_code, c=self.code, m=self.message))
 
+class Paginate:
```
I think this can be implemented more succinctly  with a generator (`yield`) instead of implementing the iterator protocol.

Something like this (untested):

```python

def paginate(con: "Connection", url: str, callback: Callback):
    page = 0
    while True:
        resp = con.get(url).json()
        yield callback(resp, page)
        next_links = [link for link in resp.get("links", []) if link["rel"] == "next] 
        if not next_links:
             break
        url = next_links[0]
        page += 1
        


```

m-mohr replied:
Does that return? I guess I need to learn what yield is before...

m-mohr replied:
How would the method be used then? Is it the same as before?

soxofaan replied:
yes both are basically equivalent,
 `yield` allows you to create an (lazy) iterator in a more to-the-point way than the iterator protocol (`__iter__` and `_next__`)

m-mohr replied:
Implemented, works, thanks.

soxofaan commented on openeo/rest/connection.py:
```diff
@@ -588,6 +617,30 @@ def describe_collection(self, name) -> dict:
+            datetimes = map(lambda t: '..' if t is None else rfc3339.normalize(t), temporal_extent)
+            q.append('datetime={}'.format('/'.join(datetimes)))
+        
+        url = '/collections/{cid}/items?{q}'.format(cid = name, q = '&'.join(q))
```
I think we can avoid all this query parameter serialization, the `requests` library will take care of that for you. e.g

```python
params = {}
if spatial_extent:
    params["bbox"] = ",".join(str(c) for c in spatial_extent)
if temporal_extent:
    params["datetime"] = "/".join(".." if t is None else rfc3339.normalize(t) for t in temporal_extent)

con.get(url, params=params)
```

m-mohr replied:
Great, didn't know...

soxofaan reviewed (commented):
added some initial review notes

m-mohr commented on openeo/rest/connection.py:
```diff
@@ -588,6 +617,30 @@ def describe_collection(self, name) -> dict:
         data = self.get('/collections/{}'.format(name)).json()
         return VisualDict("collection", data = data)
 
+    def collection_items(self, name, spatial_extent: Optional[Dict[str, float]] = None, temporal_extent: Optional[List[Union[str, datetime.datetime]]] = None) -> Paginate:
```
What is the return type? I can't figure it out with my limited Python knowledge

m-mohr replied:
How would the method be used then? Is it the same as before?

soxofaan replied:
In this case it will be an iterator of `VisualDict` if I understand correctly, so `typing.Iterator[VisualDict]`

m-mohr replied:
Took "Iterator" for now...

m-mohr replied:
With recent changes it will be more generally Iterator[dict], thanks.

m-mohr commented on openeo/rest/connection.py:
```diff
@@ -55,6 +55,34 @@ def __init__(self, http_status_code: int = None,
 
+class Paginate:
+
+    def __init__(self, con, url: str = None, callback: Callable = None):
```
I'll remove the default for url and make it so that it doesn't require a callback (is there a better term in py?)

soxofaan commented on openeo/internal/jupyter.py:
```diff
@@ -1,16 +1,89 @@
     'file-format': 'format',
     'file-formats': 'formats',
+    'item': 'data',
+    'items': 'items',
```
"item"/"items" sounds very generic, can't we make it a bit more concrete and self-describing, like "stac item" or "collection item"? Or is it really generic/abstract at this point?

m-mohr replied:
This is a mapping of component and parameter for the Vue component. This directly translates into code, we can't change anything here.

soxofaan replied:
> No user will see it.

I'm not sure, if they want to work with `collection_items` results outside of a jupyter context, they;'ll will have to understand what  `VisualDict("items"` is about

m-mohr replied:
Maybe, but as I said, that's the API of the vue-components and there's nothing we can change about it (except adding a comment behind the line). On the other hand, VisualDict is just a dict?!

m-mohr replied:
I just realized I can remove the line completely.

m-mohr commented:
Except for the "feature checking" (not sure yet how to handle it), I think I've addressed all feedback. Could you have a look again, please? @soxofaan

soxofaan commented on openeo/rest/connection.py:
```diff
@@ -55,6 +55,21 @@ def __init__(self, http_status_code: int = None,
+    page = 1
+    while True:
+        response = con.get(url, params = params).json()
+        if callback:
```
even easier than `if callback ... else` : give callback this default value: `lambda resp, page: resp`

The has added value that the `callback` argument is more self-documenting

m-mohr replied:
Sure, changed.

soxofaan commented on openeo/rest/connection.py:
```diff
@@ -55,6 +55,21 @@ def __init__(self, http_status_code: int = None,
+             break
+        url = next_link['href']
+        page += 1
+        params = {}
```
Are you intentionally clearing the `params`?  Won't that change the result set , and consequently the paging?

m-mohr replied:
Yes, the params must be in the url we get back two lines above.

soxofaan commented on openeo/rest/job.py:
```diff
@@ -264,6 +265,10 @@ def __init__(self, job: RESTJob):
         return "<JobResults for job {j!r}>".format(j=self._job.job_id)
 
+    def _repr_html_(self):
+        response = self.get_metadata()
```
This is tricky:  `get_metadata()` will (currently) fail when the job is not finished yet (`requests to this endpoint MUST be rejected with openEO error JobNotFinished`) , you don't want this behavious in `_repr_html_`

m-mohr replied:
So I guess I should just catch the error and print something else? Or what would be better?

soxofaan replied:
I think I can improve the exception that will be thrown there, so that you can check HTTP status code and openeo Error code, then you can indeed render some kind of "job not finished yet" message

soxofaan replied:
aha probably no change required: you should already get a `OpenEoApiError` exception there with `.http_status_code` >= 400 and `.code` == "JobNotFinished"

m-mohr replied:
I've tried it, but I can't import OpenEoApiError in Job, seems like a recursive dependency. How to solve that?

m-mohr replied:
Shall I just check for OpenEoClientException instead?

m-mohr replied:
...but then I can't access the message

m-mohr replied:
Implemented, that was more complicated than I thought it would be...

soxofaan commented:
Can you also add a CHANGELOG entry?

soxofaan commented on openeo/rest/job.py:
```diff
@@ -264,6 +265,13 @@ def __init__(self, job: RESTJob):
+            response = self.get_metadata()
+            return render_component("batch-job-result", data = response)
+        except OpenEoApiError as error:
+            return render_error(error, "Job not finished yet")
```
There could also be other errors, I would only catch the `JobNotFinished` ones, e.g.:

```python
try:
    response = self.get_metadata()
except OpenEoApiError as error:
    if error.code == "JobNotFinished":
        return render_error(error, "Job not finished yet")
    raise
else:
    return render_component("batch-job-result", data = response)


(the empty `raise` reraises the existing exception)

m-mohr replied:
That is handled, usually, it should just show the message. The message I'm passing there is just a fallback (maybe we don't even need the fallback?)

m-mohr replied:
Yeah, I think I'll just remove the fallback.

soxofaan replied:
my point is that you should, I think, only catch the "JobNotFinished" error, and let other issues propagate normally

m-mohr replied:
Why not show the others nicely? It's just errors that really come from the API, usually with a message, right? I'm not catching all errors, I think.

soxofaan replied:
ok, true. It's a bit a gray zone: the user does not call `_repr_html_` knowingly, so maybe it's fine to intercept exceptions a bit more greedily.

m-mohr commented:
I've added changelog entries.

soxofaan commented on CHANGELOG.md:
```diff
@@ -7,13 +7,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
 ### Added
+- `collection_items` to request collection items through a STAC API
```
hmm, maybe you should rebase on latest master (or merge master), there should be already "added" entries here

m-mohr replied:
Done

soxofaan commented on openeo/rest/connection.py:
```diff
@@ -588,6 +587,30 @@ def describe_collection(self, name) -> dict:
+        url = '/collections/{}/items'.format(name)
+        params = {}
+        if spatial_extent:
+            params["bbox"] = ",".join(str(c) for c in spatial_extent)
```
`spatial_extent` is typed as `Dict[str, float]` (so a bounding box dictionary I guess), but `str(c) for c in spatial_extent` will iterate over the  keys ("west", "north", "south", "east" in some undetermined order), which is not what you want I guess

m-mohr replied:
I copy-pasted that from load_collection. I guess we can also just allow float here? Otherwise, same order as in load_collection. So if there's an easy way to handle that as in load_collection, we could also just do that. But I was a bit confused about how load_collection did it...

soxofaan replied:
in `load_collection` and `filter_bbox` we want a bbox dictionary in the end, 
but if I understand `/collections/{}/items` correctly, bbox should be an array of length `2*n`, which is totally different

m-mohr replied:
Indeed. I committed a new version (list instead dict + removing str + document the order)

soxofaan replied:
deserves some unit tests too, I'd say :)

m-mohr replied:
https://github.com/Open-EO/openeo-python-client/issues/186

soxofaan commented on openeo/rest/connection.py:
```diff
@@ -41,20 +41,18 @@ def url_join(root_url: str, path: str):
-        self.url = url
-        super().__init__("[{s}] {c}: {m}".format(s=self.http_status_code, c=self.code, m=self.message))
-
+def paginate(con, url: str, params: dict = {}, callback: Callable = lambda resp, page: resp):
```
do you have experience with unit tests and pytest @m-mohr ? I think this one deserves a bit of coverage.
If not, I can have a look after merging

m-mohr replied:
Unit Tests yes, pytest no.

soxofaan replied:
we use a http mock library (requests_mock) for fake url serving, so we don't need a real "online" endpoint

m-mohr replied:
Good.
To get things in so that others can test things for the SRR, I'd rather postpone to later. If you don't have the time, I can give it a try, maybe we can just open an issue so that we don't forget.

m-mohr replied:
https://github.com/Open-EO/openeo-python-client/issues/186

soxofaan commented:
Overall, I think I nitpicked enough :stuck_out_tongue:  and we can wrap this up. Disclaimer: I didn't test any of this for real, I just reviewed the source code.

One general remark is that we are adding yet another method (`collection_items`) method on `Connection`. I think we should avoid putting "everything but the kitchen sink" in `Connection` as this can be annoying for documentation overview and long term maintenance. It's cleaner to split off subsystems to dedicated classes (like with batch job results). However, in the case of collections we don't have a dedicated class for that, so I don't have a good alternative for how we do it in the current PR. If you agree with this: we could put a note "this is experimental API and might be subject to change" in the doc block.

clausmichele commented:
@soxofaan when do you plan to merge this? It is necessary to have it to conclude the metadata part of the notebooks, let me know if I can test it somehow.

soxofaan commented:
I can merge today, if ok for @jdries

jdries commented:
Ok for merge, go ahead!

m-mohr commented:
@soxofaan Yes, I agree. The next step would be a `download_collection_items` or something similar and - at the moment - would also go into Connection. Thus having a collection class would really be a good idea. There's already the Collection Metadata class though?! I'll add the experimental text.

m-mohr commented:
Done, ready to be merged.

soxofaan commented:
@clausmichele  it has been merged
